package org.tvheadend.tvhclient.ui.features.playback.external

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.lifecycle.ViewModelProvider
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.PlayActivityBinding
import org.tvheadend.tvhclient.ui.common.onAttach
import timber.log.Timber
import androidx.core.net.toUri
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.color.MaterialColors
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.internal.EdgeToEdgeUtils
import com.google.android.material.shape.MaterialShapeDrawable
import com.google.android.material.shape.ShapeAppearanceModel
import org.tvheadend.tvhclient.util.applyNavigationBarPadding

abstract class BasePlaybackActivity : AppCompatActivity() {

    lateinit var binding: PlayActivityBinding
    lateinit var viewModel: ExternalPlayerViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = PlayActivityBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        EdgeToEdgeUtils.applyEdgeToEdge(window, true)
        binding.content.applyNavigationBarPadding()

        binding.content.background = MaterialShapeDrawable().apply {
            val cornerSize = resources.getDimension(R.dimen.bottom_sheet_corner_size)
            shapeAppearanceModel = ShapeAppearanceModel.builder()
                .setTopLeftCornerSize(cornerSize)
                .setTopRightCornerSize(cornerSize)
                .build()
            fillColor = ColorStateList.valueOf(
                MaterialColors.getColor(binding.content, R.attr.colorSurfaceContainer)
            )
        }

        val bottomSheetBehavior = BottomSheetBehavior.from(binding.content)
        // Expanded by default
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED
        bottomSheetBehavior.skipCollapsed = true
        bottomSheetBehavior.addBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                if (newState == BottomSheetBehavior.STATE_HIDDEN) {
                    finish()
                    //Cancels animation on finish()
                    overridePendingTransition(0, 0)
                }
            }
            override fun onSlide(bottomSheet: View, slideOffset: Float) {
            }
        })

        binding.status.setText(R.string.connecting_to_server)

        viewModel = ViewModelProvider(this)[ExternalPlayerViewModel::class.java]

        viewModel.isConnected.observe(this) { isConnected ->
            if (isConnected) {
                Timber.d("Received live data, connected to server, requesting ticket")
                binding.status.setText(R.string.requesting_playback_information)
                viewModel.requestTicketFromServer(intent.extras)
            } else {
                Timber.d("Received live data, not connected to server")
                binding.progress.isVisible = false
                binding.status.setText(R.string.connection_failed)
            }
        }

        viewModel.isTicketReceived.observe(this) { isTicketReceived ->
            Timber.d("Received ticket $isTicketReceived")
            if (isTicketReceived) {
                binding.progress.isVisible = false
                binding.status.text = getString(R.string.connected_to_server)
                onTicketReceived()
            }
        }
    }

    override fun attachBaseContext(context: Context) {
        super.attachBaseContext(onAttach(context))
    }

    protected abstract fun onTicketReceived()

    internal fun startExternalPlayer(intent: Intent) {
        Timber.d("Starting external player for given intent")
        // Start playing the video in the UI thread
        this.runOnUiThread {
            Timber.d("Getting list of activities that can play the intent")
            try {
                Timber.d("Found activities, starting external player")
                startActivity(intent)
                finish()
            } catch (ex: ActivityNotFoundException) {
                Timber.d("List of available activities is empty, can't start external media player")
                binding.status.setText(R.string.no_media_player)

                // Show a confirmation dialog before deleting the recording
                MaterialAlertDialogBuilder(this@BasePlaybackActivity)
                    .setTitle(R.string.no_media_player)
                    .setMessage(R.string.show_play_store)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        try {
                            Timber.d("Opening play store to download external players")
                            val installIntent = Intent(Intent.ACTION_VIEW)
                            installIntent.data = "market://search?q=free%20video%20player&c=apps".toUri()
                            startActivity(installIntent)
                        } catch (t2: Throwable) {
                            Timber.d("Could not startPlayback google play store")
                        } finally {
                            finish()
                        }
                    }
                    .setNegativeButton(android.R.string.cancel) { _, _ -> finish() }
                    .show()
            }
        }
    }
}
